### ISSUE TITLE
Codegen emits function declaration twice with different linkages 

### ISSUE THREAD

**Issue by awson** [leanprover/lean4#1148]
### Description

I tried to compile Lean from sources using Visual C++.

When compiling stage0 stdlib the compiler spits (among others) the following 2 error messages:
```
C:\G\sources2\LEAN\lean\lean4\stage0\stdlib\Lean\Meta\ExprDefEq.c(219): error C2375: 'lean_is_expr_def_eq': redefinition; different linkage
C:\G\sources2\LEAN\lean\lean4\stage0\stdlib\Lean\Meta\ExprDefEq.c(123): note: see declaration of 'lean_is_expr_def_eq'
 ```
and
```
C:\G\sources2\LEAN\lean\lean4\stage0\stdlib\Lean\Meta\LevelDefEq.c(53): error C2375: 'lean_is_level_def_eq': redefinition; different linkage
C:\G\sources2\LEAN\lean\lean4\stage0\stdlib\Lean\Meta\LevelDefEq.c(42): note: see declaration of 'lean_is_level_def_eq'
```

And indeed, `lean_is_expr_def_eq` is first declared as
```
lean_object* lean_is_expr_def_eq(lean_object*, lean_object*, lean_object*, lean_object*, lean_object*, lean_object*, lean_object*);
```
and a bit later redeclared (in the same file) as
```
LEAN_EXPORT lean_object* lean_is_expr_def_eq(lean_object*, lean_object*, lean_object*, lean_object*, lean_object*, lean_object*, lean_object*);
```
(same for `lean_is_level_def_eq`)

### Versions

2022-05-09 GMT 05:15

**Comment by awson** [comment]
I've looked into (perhaps) relevant commits and noticed the following: the commit introducing `LEAN_EXPORT` looks as if `isExternal` is now used instead of `addExternForConsts`, [but `emitFnDeclAux` is now called with `extC` instead of `!extC`](https://github.com/leanprover/lean4/commit/35ffae6f54c59674def1ac62de23d713ff9c79eb#diff-83d471598630766cc6f751a0ab55476a7a4e2557104054d8bc631314f73b86c6L121).

Is this OK?

**Comment by Kha** [comment]
Yes, that code assumes that any `[extern]` decl is truly external, i.e. defined in a different object file. However, `Lean.Meta.ExprDefEq` both exports `lean_is_expr_def_eq` *and* uses the previously defined `@[extern 7 "lean_is_expr_def_eq"] constant isExprDefEqAux`, which confuses the C emitter. I don't see a trivial solution/workaround here (and suspect that there may be other issues with compiling Lean with VC++).

**Comment by awson** [comment]
Ah, understood.

Yes, VC++ requires other fixes, but they are more or less straightforward.

I can build stage0 lean successfully with VC++ if grepping out these 2 offending lines.

**Comment by awson** [comment]
Hmm, I've made quick experiment and found that `__declspec(dllexport)` is absolutely harmless when we declare an external.

**Comment by Kha** [comment]
That's good to hear. But I also just remembered this reservation about `dllimport`: https://github.com/leanprover/lean4/pull/670#issuecomment-920339420. stage0 might not be affected by this because no Lean code is shared across DLLs there yet.

**Comment by awson** [comment]
> In the latter case we have a small performance penalty (indirect data access)

And `mingw` target **always do this**, and this is why the can seamlessly import data from dlls (runtime pseudo-relocs). Any `extern` data (not a function) is accessed through a variable stub.

**Comment by Kha** [comment]
I see, that makes sense. If you can make this work, I'm happy to review it.

**Comment by awson** [comment]
[In your comment](https://github.com/leanprover/lean4/pull/670#issuecomment-920339420) you write

**Comment by Kha** [comment]
Yes, that sounds right
